### PR TITLE
[AssetMapper] Fix `JavaScriptImportPathCompiler` regression in regex

### LINKED ISSUE #54078: [AssetMapper]  JavaScriptImportPathCompiler::compile(): Return value must be of type string, null returned
### Symfony version(s) affected

6.4.4

### Description

Regression in AssetMapper 6.4.4 released today

ScreenShot-2024-02-27-08 57 24

### How to reproduce

```
symfony new . --version="6.4.*" --webapp
php bin/console importmap:req @yaireo/tagify
php bin/console importmap:update 
```

ScreenShot-2024-02-27-09 20 40

### Possible Solution

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by PhilETaylor**
| Q             | A
| ------------- | ---
| Branch?       | 6.4
| Bug fix?      | yes
| New feature?  | no 
| Deprecations? | no 
| Issues        | Fix #54078
| License       | MIT

Fixes regression in regex, fix provided by @nicolas-grekas in #54078

### CODE REVIEW

**Comment by PhilETaylor** [comment]
I have also updated the https://regex101.com/r/1iBAIb/1 to https://regex101.com/r/1iBAIb/2 by adding the same change there.

**Review by stof — APPROVED** [review]

File: src/Symfony/Component/AssetMapper/Compiler/JavaScriptImportPathCompiler.php [hunk 1]
```diff
@@ -28,16 +28,16 @@
 final class JavaScriptImportPathCompiler implements AssetCompilerInterface
 {
     /**
-     * @see https://regex101.com/r/1iBAIb/1
+     * @see https://regex101.com/r/1iBAIb/2
      */
     private const IMPORT_PATTERN = '/
         ^
             (?:\/\/.*)                     # Lines that start with comments
```
Comment (nicolas-grekas) [inline src/Symfony/Component/AssetMapper/Compiler/JavaScriptImportPathCompiler.php]:
```suggestion
            ^(?:\/\/.*)                     # Lines that start with comments
```

Comment (PhilETaylor) [reply src/Symfony/Component/AssetMapper/Compiler/JavaScriptImportPathCompiler.php]: dammit, I suck. Thanks for checking my work!

**Comment by weaverryan** [comment]
Thanks for this :). Do we know if there is a test case that could be created to show the problem? Due to the complexity of this regex, we rely heavily on the tests to avoid regressions like this.

**Comment by PhilETaylor** [comment]
@weaverryan The reproducer I created was in the linked issue https://github.com/symfony/symfony/issues/54078

```
symfony new . --version="6.4.*" --webapp
php bin/console importmap:req @yaireo/tagify
php bin/console importmap:update 
```

**Comment by PhilETaylor** [comment]
If Im honest I dont really know what the root problem is enough to create some regression unit tests sorry. I just stumbled on and reported the issue first once I had a reproducer, and used @nicolas-grekas provided solution as a PR when requested.

**Comment by nicolas-grekas** [comment]
This happens because of big js files to parse with some content that trigger deep backtracking. Not easy to reliably test I fear...

**Comment by smnandre** [comment]
Thank you @PhilETaylor for this fix. 
We're trying to handle so many cases with a single regexp... 

Hopefully this one would bring us near the 99% we target, without having to implement a real JS parser (that would slow things down **a lot** for everyone)

**Comment by smnandre** [comment]
> This happens because of big js files to parse with some content that trigger deep backtracking. Not easy to reliably test I fear...

I experienced the bug right now, while updating a branch i was working on the ux website... Fixed locally with a dirty patch...

```diff
            $relativeImportPath = $this->makeRelativeForJavaScript($relativeImportPath);

            return str_replace($importedModule, $relativeImportPath, $fullImportString);
-        }, $content, -1, $count, \PREG_OFFSET_CAPTURE);
+        }, $content, -1, $count, \PREG_OFFSET_CAPTURE) ?? $content;
    }
```    

With your example i think i see the problem here ...  https://regex101.com/r/WueQkU/1 

Capture d’écran 2024-02-29 à 14 00 57

.. and this file does not contains a single "`import`" 😭 

So what about we also add a pre-check and ignore files that do not contain "`import`" ?

**Comment by stof** [comment]
@smnandre A `str_contains` check for `import` might indeed help for the case of files without any imports to exclude them faster. But I think we could still have such nasty backtracking for big JS files containing. So we need this PR.

Btw, I just thought about a case not covered by the current regex. But I'll wait for this PR to be merged before doing my PR as they would conflict.

**Comment by smnandre** [comment]
> So we need this PR.

Of course! I just suggested an _additional_ check :)

I'll open another one to throw an expression instead of returning null.

**Comment by nicolas-grekas** [comment]
Thank you @PhilETaylor.

**Comment by ph-il** [comment]
Got the same problem with 
'@fullcalendar/core/internal.js' => [
    'version' => '6.1.11',
], 
So if you need more test case.

**Pull request merged**